requests/utils.py

```diff
@@ -90,6 +86,19 @@ def super_len(o):
             # let requests chunk it instead.
             current_position = total_length
 
+    if hasattr(o, 'seek') and total_length is None:
+        # StringIO has a notimplemented fileno
+        # BytesIO has seek and not fileno
+        total_length = o.seek(0, 2)
+        # StringIO objects in py2 returns None
+        if total_length is None and hasattr(o, 'tell'):
+            total_length = o.tell()
+        elif total_length is None:
+            total_length = 0
+
+        # seek back to current position to support partially read file-like
+        o.seek(current_position or 0)
+
     return max(0, total_length - current_position)
```

Comment:
So with this patch we can fall down to here with no `total_length` and then do some arithmetic on it, which will blow up. We need a way to default `total_length` to 0 right before this. The same problem applies to `current_position`, potentially.